File: apps/windows/TeslaSync.App.Core/Push/PushPayloadParser.cs
using System.Text.Json;

namespace TeslaSync.App.Core.Push;

/// <summary>
/// Decodes a raw WNS foreground push body into a typed <see cref="PushPayload"/> (P2/W6-0002). The
/// parser is tolerant: an unparseable or non-object body yields <see cref="PushPayload.Unknown"/>
/// rather than throwing, so a malformed push can never crash the foreground pump. It mirrors the
/// backend <c>notification-worker</c> envelope (a JSON object with <c>kind</c>/<c>type</c>,
/// <c>title</c>, <c>body</c>/<c>message</c> and <c>category</c> plus arbitrary string <c>data</c>).
/// </summary>
public static class PushPayloadParser
{
    /// <summary>Parses <paramref name="raw"/>; returns <see cref="PushPayload.Unknown"/> on any failure.</summary>
    public static PushPayload Parse(string? raw)
    {
        if (string.IsNullOrWhiteSpace(raw))
        {
            return PushPayload.Unknown;
        }

        try
        {
            using var document = JsonDocument.Parse(raw);
            var root = document.RootElement;
            if (root.ValueKind != JsonValueKind.Object)
            {
                return PushPayload.Unknown;
            }

            var kind = ReadString(root, "kind") ?? ReadString(root, "type") ?? PushPayload.UnknownKind;
            var title = ReadString(root, "title");
            var body = ReadString(root, "body") ?? ReadString(root, "message");
            var category = ReadString(root, "category");
            var data = ReadData(root);

            return new PushPayload(kind, title, body, category, data);
        }
        catch (JsonException)
        {
            return PushPayload.Unknown;
        }
    }

    private static string? ReadString(JsonElement element, string name) =>
        element.TryGetProperty(name, out var value) && value.ValueKind == JsonValueKind.String
            ? value.GetString()
            : null;

    private static Dictionary<string, string> ReadData(JsonElement root)
    {
        var data = new Dictionary<string, string>(StringComparer.Ordinal);
        if (root.TryGetProperty("data", out var element) && element.ValueKind == JsonValueKind.Object)
        {
            foreach (var property in element.EnumerateObject())
            {
                if (property.Value.ValueKind == JsonValueKind.String)
                {
                    data[property.Name] = property.Value.GetString() ?? string.Empty;
                }
            }
        }

        return data;
    }
}

File: apps/android/app/src/main/java/io/teslasync/android/push/PushPayloadParser.kt
package io.teslasync.android.push

/**
 * Decodes an FCM message into a typed [PushPayload] (P3/A6). FCM already delivers the data part as a
 * `Map<String, String>`, so no JSON parsing is needed; the optional notification title/body (present
 * on notification-style messages) are folded in as fallbacks. The parser is tolerant: an empty
 * message yields [PushPayload.Unknown] so a malformed push can never crash the foreground pump. It
 * mirrors the backend `notification-worker` envelope (`kind`/`type`, `title`, `body`/`message`,
 * `category` plus arbitrary string `data`).
 */
object PushPayloadParser {
    /** Parses the FCM [data] map plus any notification title/body; returns [PushPayload.Unknown] when empty. */
    fun parse(
        data: Map<String, String>,
        notificationTitle: String? = null,
        notificationBody: String? = null,
    ): PushPayload {
        if (data.isEmpty() && notificationTitle == null && notificationBody == null) {
            return PushPayload.Unknown
        }
        val kind = data["kind"] ?: data["type"] ?: PushPayload.UNKNOWN_KIND
        val title = data["title"] ?: notificationTitle
        val body = data["body"] ?: data["message"] ?: notificationBody
        return PushPayload(kind, title, body, data["category"], data)
    }
}
